onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc

```diff
@@ -20,54 +23,108 @@ bool CanNodePropagate(const Node& node) {
+
+    const Node* edge_dst_node = insertion_edge.GetNodeAtEnd(graph, ExtendedGraphEdge::End::Destination);
+    ORT_RETURN_IF_NOT(src_node != nullptr || edge_dst_node != nullptr,
+                      "At least one graph node must be specified in the propagation edges.");
+  }
+
+  return Status::OK();
+}
+
+// Logs information about the edges into which Q/DQ nodes will be inserted in InsertQDQPairs().
+// Assumes the edges have already been validated.
+void LogQDQInsertion(const logging::Logger& logger, logging::Severity severity,
+                     const Graph& graph, const InlinedVector<ExtendedGraphEdge>& edges) {
+  if (!logger.OutputIsEnabled(severity, logging::DataType::SYSTEM)) {
+    return;
+  }
+
+  const Node* src_node = edges[0].GetNodeAtEnd(graph, ExtendedGraphEdge::End::Source);
+  const auto& node_arg_name = edges[0].arg_name;
+  std::string src_label = src_node ? MakeString("node (\"", src_node->Name(), "\", index: ", src_node->Index(), ")")
+                                   : "input";
+  std::ostringstream dst_labels;
+  const size_t num_edges = edges.size();
+
+  for (size_t i = 0; i < num_edges; ++i) {
+    const ExtendedGraphEdge& edge = edges[i];
+    const Node* dst_node = edge.GetNodeAtEnd(graph, ExtendedGraphEdge::End::Destination);
+    dst_labels << (dst_node ? MakeString("dst node (\"", dst_node->Name(), "\", index: ", dst_node->Index(), ")")
+                            : "output")
+               << (i == num_edges - 1 ? "" : ",");
+  }
 
   LOGS(logger, VERBOSE) << "Inserting Q/DQ pair between "
-                        << (src_node ? MakeString("node (\"", src_node->Name(), "\", index: ", src_node->Index(), ")")
+                        << (src_node ? MakeString("src node (\"", src_node->Name(), "\", index: ", src_node->Index(), ")")
                                      : "input")
-                        << " and "
-                        << (dst_node ? MakeString("node (\"", dst_node->Name(), "\", index: ", dst_node->Index(), ")")
-                                     : "output")
-                        << " at NodeArg \"" << base_name << "\".";
+                        << " and " << dst_labels.str()
+                        << " at NodeArg \"" << node_arg_name << "\".";
+}
+
+// convert this: src_node --+--> dst_node_0
+//                          |
+//                          +--> dst_node_1
+//                          |    ...
+//                          +--> dst_node_n
+//
+// to this: src_node -> Q --+--> DQ -> dst_node_0
+//                          |
+//                          +--> DQ -> dst_node_1
+//                          |    ...
+//                          +--> DQ -> dst_node_n
+// assumptions:
+// 1. All insertion edges have the same source node and the same source node output index.
+// 2. Insertion_edges are valid: node indices refer to valid nodes, and arg names refer to valid NodeArgs in the graph.
```

Comment:
for 1., all edge sources should be either the same node output or the same graph input/initializer

for 2., is it the case that there should only be a single NodeArg that is referred to by all the edges?

Reply:
I updated the comments to note that all edges do share the same NodeArg (come from either the same source node's output or they come from the same graph input). The function returns an error status if this is not the case.